---
title: "BCB420H Assignment 2"
author: "Terrence Amponsah"
output:
  html_document:
    toc: true
student_no: 1005413429
bibliography: a2.bib
---

# 1. Intro:
## Setup: 
### R:
Installing the R packages used in this notebook.
```{r, message=FALSE}
if (!requireNamespace(c("BiocManager", "GEOmetadb", "knitr", "edgeR", "limma"), quietly = TRUE)){
  install.packages("BiocManager")
  install.packages("GEOmetadb")
  install.packages("knitr")
  install.packages("edgeR")
  install.packages("limma")
}



library(DBI)
library(GEOmetadb)
library(GEOquery)
library(edgeR)
BiocManager::install("biomaRt")
library(limma)

# A2 it seems.
library (ComplexHeatmap)
library(circlize)
```

###  Dataset: 
INTRO ABOUT DATASET!!!!
For analysis, I chose the dataset [GSE126848](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126848)[@Suppli2019-hz]

### [Dataset](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE126848) setup:
```{r, message=FALSE}

suppFiles = GEOquery::getGEOSuppFiles('GSE126848')
fileNames = rownames(suppFiles)

nafld_exp = read.delim(fileNames[1], header=TRUE, check.names=FALSE)
gse <- GEOquery::getGEO("GSE126848", GSEMatrix = FALSE)
current_gpl <-names(GEOquery::GPLList(gse))[1]
current_gpl_info <-GEOquery::Meta(GEOquery::getGEO(current_gpl))
```

### Info about my dataset: 

* **Platform:** `r current_gpl_info$title`  
* **Submission date:** `r current_gpl_info$submission_date`  
* **Last update date:** `r current_gpl_info$last_update_date`   
* **Organism(s):** `r current_gpl_info$organism`   
* **\# Geo datasets which use this technology:** `r length(current_gpl_info$series_id)`   
* **\# Geo samples that use this technology:** `r length(current_gpl_info$sample_id)`

***


I made this samples data frame to use in my A1 data, allows a mapping between 
the ids and the patient numbers, ids could be collected from GEO but, not sure 
if it is explicitly stored.
```{r}
patient_no <- as.character(c(1:15, 1:16, 1:14, 1:12))
group <- rep(c("normal_weight", "NASH", "NAFL", "Obese"), times=c(15,16,14,12))
ids <- c("0869", "0873", "0875", "0877", "0879", "0881", "0883", "0885", "0887",
         "0889", "0891", "0893", "0897", "0910", "2684", "2688", "2692", "2696", 
         "2698", "2704", "2705", "3992", "3994", "3996", "3997", "4000", "4004", 
         "4005", "4006", "4008", "2683", "2685", "2687", "2689", "2691", "2693", 
         "2697", "2701", "2703", "3993", "3995", "3998", "4002", "4007", "4010", 
         "0872", "0874", "0876", "0878", "0886", "0888","0890", "0892", "0894", 
         "0896", "0898", "0899")
samples <- data.frame(patient_no=patient_no, group=group, ids=ids)
```

The sample dataset from the GEO database does not have any HUGO identifiers, so we need to add this to the data: 
### Adding HUGO to dataset:
```{r}
nafld_expc <- nafld_exp

ensembl = biomaRt::useMart("ensembl", dataset = "hsapiens_gene_ensembl") # takes time
gene <- biomaRt::getBM(attributes = c("ensembl_gene_id", "hgnc_symbol"), 
              values = nafld_exp$key, mart=ensembl)
id <- match(nafld_exp$key, gene$ensembl_gene_id)
nafld_exp$HUGO_id <- gene$hgnc_symbol[id]
```

### Cleaning the data: 
```{r}
cpms = edgeR::cpm(nafld_exp[,2:58])

# Using n = 12 (loosest i think). 
# The publication has 4 n values (12, 14, 15, 16)
# One for each category $\in$ {"obese", "NASH", "normal-weight", "NADL"}
keep = rowSums(cpms > 1 ) >= 12
filtered_ds = nafld_exp[keep,]

summarized_gene_counts_filtered <- sort(table(filtered_ds$HUGO_id),
                                        decreasing = TRUE)
```
***

### Normalization(TMM):
```{r}
filtered_ds_mtx <- as.matrix(filtered_ds[,2:58])
rownames(filtered_ds_mtx) <- filtered_ds[,59] #proper names this time around.

dge <- edgeR::DGEList(counts = filtered_ds_mtx, group=samples$group)

norm_fact <- edgeR::calcNormFactors(dge)
normalized_counts <- edgeR::cpm(norm_fact)
knitr::kable(head(normalized_counts), output="html")
```

###ID Mapping with Ensembl
```{r, results='hide'}
conversion_stash <- "nafld_id_conversion.rds"

if(file.exists(conversion_stash)){
  nafld_id_conversion <- readRDS(conversion_stash)
} else {
  nafld_id_conversion <- getBM(
                            attributes = c("ensembl_gene_id", "hgnc_symbol"),
                            filters = c("ensembl_gene_id"), 
                            values = filtered_ds$key, 
                            mart = ensembl
  )
}
nc_annot <- merge(nafld_id_conversion, normalized_counts, by.x = 2, by.y = 0, all.y = TRUE)

missing_ids <- nc_annot$hgnc_symbol[which(is.na(nc_annot$hgnc_symbol) | is.na(nc_annot$ensembl_gene_id))]

mid_ds <- data.frame(gsub("\\.", "-", missing_ids))
colnames(mid_ds) <- c("missing_ids")

old_mapping <- merge(nafld_exp[,c(1,ncol(nafld_exp))], 
                     mid_ds, by.x="HUGO_id", by.y = "missing_ids")
dots_to_dashes <- data.frame(c(data.frame(gsub('\\.', '-', missing_ids), missing_ids)))

dot_hgnc <- merge(old_mapping, dots_to_dashes, by.x = 0, by.y = 0)
dot_hgnc <- dot_hgnc[c("key", "missing_ids")]

mis <- nc_annot[which(is.na(nc_annot$ensembl_gene_id) | is.na(nc_annot$hgnc_symbol)),]
mis_withids <- merge(dot_hgnc, mis, by.x = "missing_ids", by.y = "hgnc_symbol")
colnames(mis_withids)[0:1] <- colnames(nc_annot)[0:1]

finalized_ncc <- rbind(nc_annot[which(
  !is.na(nc_annot$ensembl_gene_id) | !is.na(nc_annot$hgnc_symbol)),], mis_withids[, colnames(nc_annot)])

dup_ens <- finalized_ncc$ensembl_gene_id[which(duplicated(finalized_ncc$ensembl_gene_id))]
nodup_f_nc <- finalized_ncc[-which(duplicated(finalized_ncc$ensembl_gene_id)),]

final_ncc <- nodup_f_nc[-which(is.na(nodup_f_nc$ensembl_gene_id)),]
```


## HeatMap Prep: 


### Normalized counts to numerical matrix: 
```{r}
knitr::kable(final_ncc[1:5,1:5], type="html")

num_matrix <- final_ncc[,3:ncol(final_ncc)]
rownames(num_matrix) <- final_ncc$ensembl_gene_id
colnames(num_matrix) <- colnames(final_ncc[, 3:ncol(final_ncc)])
```
now that the data is in matrix form, I a Heatmap can be created.

<!-- ## Heatmap (no Row-normalization):
```{r, eval=FALSE}
if(min(num_matrix) == 0) {
  heatmap_col = circlize::colorRamp2(c(0, max(num_matrix)), 
                          c( "white", "red"))
} else {
  heatmap_col = circlize::colorRamp2(c(min(num_matrix), 0, max(num_matrix)), 
                           c("blue", "white", "red"))
}

current_hm <- ComplexHeatmap::Heatmap(as.matrix(num_matrix),
                    )
current_hm
```
 -->
 
### Heatmap (with row-normalization)
```{r warning=FALSE}
normalized_matrix <- t(scale(t(num_matrix)))

if(min(normalized_matrix) == 0) {
  heatmap_col = circlize::colorRamp2(c(0, max(normalized_matrix)), 
                          c( "white", "red"))
} else {
  heatmap_col = circlize::colorRamp2(c(min(normalized_matrix), 0, max(normalized_matrix)), 
                           c("blue", "white", "red"))
}

current_hm <- ComplexHeatmap::Heatmap(as.matrix(normalized_matrix),
                    show_row_dend = TRUE, show_column_dend = TRUE, 
                    col = heatmap_col, show_column_names = TRUE, 
                    show_row_names = FALSE, show_heatmap_legend = TRUE)
current_hm

```

A list of genes of interest of varying function were collected from the paper. 
### Sample genes of interest collected from paper.
```{r}
nw_samples <- samples$ids[which(samples$group == "normal_weight")]
nash_samples <- samples$ids[which(samples$group == "NASH")]
ob_samples <- samples$ids[which(samples$group == "Obese")]
nafl_samples <- samples$ids[which(samples$group == "NAFL")]

goi_list <- c("HMGCS2",  "LDLR",  "SCD1", "RS2",  "G6PASE", "PPARA", "TGFB", "A-SMA",  "COL1A1",  "COL1A2","COL3A1",  "COL6A1",  "COL6A2",  "PDGF", "AEBP1", "DPT", "EFEMP1", "FBLN5","ITGBL1",  "LOXL4",  "THBS2",  "VCAN", "MAT1A", "ANTRX1", "CLDN11", "EPCAM", "MGP", "PODN","STMN2")
goi <- which(final_ncc$hgnc_symbol == goi_list[1])

(goi_nw_samples <- t(final_ncc[goi,nw_samples]))
colnames(goi_nw_samples) <- "normal_weight_samples"
goi_nash_samples <- t(final_ncc[goi,nash_samples])
colnames(goi_nash_samples) <- "NASH_samples"
goi_ob_samples <- final_ncc[goi,ob_samples]
colnames(goi_ob_samples) <- "Obese_samples"
goi_nafl_sample <- final_ncc[goi,nafl_samples]
colnames(goi_nafl_sample) <- "NAFL_samples"
```

### T.test between normal weight and NAFL.
```{r}
t.test(x=t(goi_nw_samples), y=t(goi_nafl_sample))
```

--- 
### MDS Plot for samples.
```{r}

limma::plotMDS(num_matrix,
               col = c("darkgreen", "blue", "orange", "pink")[factor(samples$group)])
```

## Model Creation 

Model design based on the "healthy", "obese", "NASH", "NAFL" groups.
```{r}
model_design <- model.matrix( ~ samples$group)
knitr::kable(model_design[1:5,], type="html")

expressionMatrix <- as.matrix(final_ncc[,3:ncol(final_ncc)])
rownames(expressionMatrix) <- final_ncc$ensembl_gene_id
colnames(expressionMatrix) <- colnames(final_ncc)[3:ncol(final_ncc)]

minimalSet <- Biobase::ExpressionSet(assayData=expressionMatrix)

fit <- limma::lmFit(minimalSet, model_design)

fit2 <- limma::eBayes(fit, trend=TRUE)

topfit <- limma::topTable(fit2, 
                  coef = ncol(model_design), 
                  adjust.method= "BH", 
                  number = nrow(expressionMatrix))

output_hits <- merge(final_ncc[,1:2], 
                     topfit, 
                     by.x = "ensembl_gene_id", by.y = "row.names", all.y = TRUE)
# REMEMBER TO USE NAMES COLUMNS, NUMERICAL ONES DON'T WORK FOR SOME REASON!

output_hits <- output_hits[order(output_hits$P.Value),]

knitr::kable(output_hits[1:10,], type="html",row.names=FALSE)
```

### Genes with p-values below threshold (0.05)
```{r} 
length(which(output_hits$P.Value < 0.05)) # No. of genes with p-value < 0.05
length(which(output_hits$adj.P.Val < 0.05)) # No. of genes which _ 
```

## Model #2, (group (from model 1) + patient_no) 
Patient
```{r}
mdp_pat <- model.matrix( ~samples$patient_no + samples$group)
fit_pat <- limma::lmFit(minimalSet, mdp_pat)
fit2_pat <- limma::eBayes(fit_pat, trend=TRUE)
topfit_pat <- limma::topTable(fit2_pat,
                       coef=ncol(mdp_pat), 
                       adjust.method = "BH", 
                       number = nrow(expressionMatrix))
output_hit_pat <- merge(final_ncc[,1:2], 
                     topfit_pat, 
                     by.x = "ensembl_gene_id", by.y = "row.names", all.y = TRUE)

output_hit_pat <- output_hit_pat[order(output_hit_pat$P.Value),]
knitr::kable(output_hit_pat[1:10, 2:8], type="html", row.names = FALSE)
```

### Up-regulated Genes: 
```{r}
length(which(output_hit_pat$P.Value < 0.05 & output_hit_pat$logFC > 0))
```

### Down-regulated Genes:
```{r}
length(which(output_hit_pat$P.Value < 0.05 & output_hit_pat$logFC < 0))
```

### P-values for model #2 : 
```{r fig.cap= "In both cases, ~ 100 genes more pass the threshold p-value and correction when accounting for patients."}
length(which(output_hit_pat$P.Value < 0.05))
length(which(output_hit_pat$adj.P.Val < 0.05))
```

## P-value plots: 
```{r}
simple_model_pvals <- data.frame(ensembl_id = 
                                   output_hits$ensembl_gene_id, 
                                 simple_pvals = output_hits$P.Value)

pm_pvals <- data.frame(ensembl_id = 
                                   output_hit_pat$ensembl_gene_id, 
                                 pat_pvals = output_hit_pat$P.Value)

two_models_pvals <- merge(simple_model_pvals, pm_pvals, by.x=1, by.y=1)
two_models_pvals$color <- "black"
two_models_pvals$color[two_models_pvals$simple_pvals<0.05] <- "orange"
two_models_pvals$color[two_models_pvals$pat_pvals<0.05] <- "blue"
two_models_pvals$color[two_models_pvals$simple<0.05 & two_models_pvals$pat_pvals<0.05] <- "red"

plot(two_models_pvals$simple_pvals, two_models_pvals$pat_pvals, col = two_models_pvals$color, 
     xlab = "Simple model p-vals", ylab = "Patient model p-vals", main="Simple v. Patient Limma model P-values")

```
### GOI highlighted
```{r}
goi_ens <- final_ncc$ensembl_gene_id[which(final_ncc$hgnc_symbol %in% goi_list)]

tm_pvals <- merge(simple_model_pvals, pm_pvals, by.x=1, by.y=1)
tm_pvals$color <- "darkgrey"
tm_pvals$color[tm_pvals$ensembl_id %in% goi_ens] <- "orange"

plot(tm_pvals$simple_pvals, tm_pvals$pat_pvals, 
     col = tm_pvals$color, xlab = "Simple model p-vals", ylab = "Patient model p-vals", main="Simple v. Patient Limma model P-values") + abline(h = 0.5, v = 0.5, lty=2)
points(tm_pvals[which(tm_pvals$ensembl_id %in% goi_ens), 2:3], pch=20, col="orange", cex=1.5)
which(final_ncc$hgnc_symbol %in% goi_list)
```

## Heatmap of the Top Hits.
```{r fig.cap = "Heatmap for __ filtered for p.value < 0.01 and ordered by group (normal_weight, obese, nash, nafl)"}
tophits_ms <- output_hit_pat$ensembl_gene_id[output_hit_pat$P.Value<0.01]
ms_mtx <- t(scale(t(num_matrix[which(rownames(num_matrix)%in% tophits_ms),])))

ms_mtx <- ms_mtx[, c(which(colnames(ms_mtx) %in% samples$ids[which(samples$group == "normal_weight")]), 
                     which(colnames(ms_mtx) %in% samples$ids[which(samples$group == "Obese")]), 
                     which(colnames(ms_mtx) %in% samples$ids[which(samples$group == "NASH")]), 
                     which(colnames(ms_mtx) %in% samples$ids[which(samples$group == "NAFL")]))]

if(min(ms_mtx) == 0) {
  heatmap_colo = circlize::colorRamp2(c(0, max(ms_mtx)), 
                          c( "white", "red"))
} else {
  heatmap_colo = circlize::colorRamp2(c(min(ms_mtx), 0, max(ms_mtx)), 
                           c("blue", "white", "red"))
}

tophits_heatmap_cl <- ComplexHeatmap::Heatmap(as.matrix(ms_mtx), 
                           cluster_rows = TRUE, cluster_columns = FALSE, 
                           show_row_dend = TRUE, show_column_dend = TRUE, 
                           col = heatmap_colo, show_column_names = TRUE, 
                           show_row_names = FALSE, show_heatmap_legend = TRUE)
tophits_heatmap_cl
```

## Thresholded List of genes: 
```{r, eval=FALSE}
ohp_named <- merge(nafld_exp[,c(1,ncol(nafld_exp))], output_hit_pat, by.x="key", by.y="ensembl_gene_id")
ohp_named[,"rank"] <- -log(ohp_named$P.Value, base=10) * sign(ohp_named$logFC)
ohp_named <- ohp_named[order(ohp_named$rank), ]

urg <- ohp_named$hgnc_symbol[which(ohp_named$P.Value < 0.05 & ohp_named$logFC > 0)]
drg <- ohp_named$hgnc_symbol[which(ohp_named$P.Value < 0.05 & ohp_named$logFC < 0)]

write.table(x=c(urg, drg), 
            file=file.path('./data', 'nafld_reg_g.txt'), sep = "\t", 
            row.names = FALSE, col.names = FALSE, quote=FALSE)

write.table(x=urg, 
            file=file.path('./data', 'nafld_urg.txt'), sep="\t", 
            row.names = FALSE, col.names = FALSE, quote = FALSE)

write.table(x = drg, file=file.path('./data', 'nafld_drg.txt') ,  sep="\t",
            row.names=FALSE, col.names = FALSE, quote = FALSE)

write.table(x = data.frame(genename=ohp_named$hgnc_symbol, F_stat=ohp_named$rank),
            file=file.path('./data', 'rankedGL.txt'), sep="\t", 
            row.names = FALSE, col.names = FALSE, quote=FALSE)
```
# Answers: 
## Differential Gene Expression:
1. & 2. Here are the number of genes which were significantly differentially expressed & the number of genes which pass correction.  The standard threshold p-value of 0.05 was used to categorize a genes a "significantly" differentially expressed.
```{r, fig.cap="The number of genes which were significantly differentially expressed according the the first model (group)", collapse=TRUE} 
length(which(output_hits$P.Value < 0.05)) # No. of genes with p-value < 0.05
length(which(output_hits$adj.P.Val < 0.05)) # No. of genes which _ 
```

```{r fig.cap= "In both cases, ~ 100 genes more pass the threshold p-value and correction when accounting for patients."}
length(which(output_hit_pat$P.Value < 0.05)) # No. of genes with p-value < 0.05
length(which(output_hit_pat$adj.P.Val < 0.05)) # No. of genes which pass correction
```

3. MA Plot of Example NAFL vs Normal-weight:
```{r}
limma::plotMA(log2(nafld_expc[,c(2,ncol(nafld_expc))]), ylab="M - ratio log expression", main="NAFL vs Normal-Weight'ed human sample comparison")
```

## Thresholded ORA
1. [G:Profiler](https://biit.cs.ut.ee/gprofiler/gost)[@Raudvere2019-od] was used for the thresholded ORA. G:Profiler was used because it has support for both gene identifiers that were used and produced results when the query was run.

2. Gene Ontology (MF, CC, BP) (release: 2021-12-15), Reactome (class: 2022-1-3) and Human Protein Atlas(script: 21-12-17) were used for annotation. These sets have information which may provide insights for my dataset.

3. After running the complete set of up- and down-regulated genes with G:Profiler, `r 90+855+212+35+165+49` genesets are returned. User Threshold is 0.05 (with BHFDR). Results were also filtered for term sizes between 3 and 1000.

4. When re-running this analysis with the up-regulated genes, the top results are very different with the down-regulated set compared to the up-regulated set, which has some returned genesets in common (e.g. GO:MF-transcription corepressor activity)

## Interpretation

1 & 2. The over-representation results seem generally in-line with the information from the paper. In figure two of [@Suppli2019-hz], the authors discuss the gene expression of transcription factors which is seen as some of the datasets (transcription corepressor/coregulator, chromatin binding, histone binding, etc...) that are retured by G:Profiler.
# References
* GSEA
* Cytoscape
* EnrichmentMap
* Dataset